sopel/plugins/capabilities.py

```diff
@@ -0,0 +1,312 @@
+"""Capability Requests management for plugins."""
+from __future__ import annotations
+
+import logging
+from typing import (
+    Dict,
+    FrozenSet,
+    Generator,
+    Iterable,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    TYPE_CHECKING,
+    Union,
+)
+
+
+if TYPE_CHECKING:
+    from sopel.bot import Sopel, SopelWrapper
+    from sopel.plugin import capability, CapabilityNegotiation
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class Manager:
+    """Manager of Capability Requests."""
+    def __init__(self):
```

Comment:
For something like this, that's designed to be instantiated as part of another object (a bot, in this case), is there anything besides the possible added difficulty of unit tests that makes you not want to make the parent object provide itself as a constructor argument, to avoid needing `bot` as method args later?

Reply:
> besides the possible added difficulty of unit tests

Not "possible": very real and very annoying. And yes there is! Beside the difficulty (more like: complexity, because state and stuff), it is already very annoying to deal with the bot's interface alone, especially when split between the `AbstractBot` base class (that handles only IRC-related stuffs) and the `Sopel` class (that handles the plugins & various setups).

Splitting that section into its own object makes it much simpler to work with, since it's simpler to reason on a very specific subset of features. This also prevents a bunch of name conflict. This is, by the way, the same reason we have a `bot.rules` object that proxy all the plugin callables management for the bot.

Reply:
Yeah, I could have asked the same question about other stuff that's been added/refactored in the last few versions, but I guess I didn't think of it.

Passing `bot` around all the time is _kind of_ annoying in the same way as always having to restate `plugin_name` (as said in that other thread here), but not as bad because it's something you're always just going to _have_ and not a string that has to be kept up-to-date if the name of something changes. (Well, OK, it _could_ change if someone's being silly with their args, but you know what I meant.)

Just to be clear, what I had in mind was (consider this as pseudo-code):

```py
# in the bot's init:
self._capability_manager = capabilities.Manager(bot=self)

# in the Manager's init
self._bot = bot

# possibly also as part of the Manager
@property
def bot(self):
    return self._bot

# then use the attr/property in Manager methods,
# instead of passing bot as an arg
```

Especially the last part of your reply sounds like you were reacting to a suggestion that this Manager be moved into the bot/Sopel class somehow, which is definitely _not_ what I meant to suggest! 😅 The modularization is good!

Reply:
Ha! I see. The thing is, in most case we are not interested in the instance of `Sopel` itself, but in a `SopelWrapper` instance, which is contextual and not available when instantiating the `Manager`. So for now, I won't change that.